### ISSUE TITLE
[Feature] Co-locate similar files when sharding

### ISSUE THREAD

**Issue by pastelsky** [microsoft/playwright#22987]
Currently, when tests are sharded, it is very likely that files that are very different parts of the application are bundled into the same shard. 

For large codebases that use component testing, this can result high bundling times. Sorting the test files by path, and then sharding will result in a much smaller module sprawl for the bundler and for large components, can have a significant impact on bundling time and hence testing run time.

For our internal use cases, where we run hundreds of tests, adding a simple `sort()` here —
https://github.com/microsoft/playwright/blob/e6d8cf9693668e4eb82360551953bda35096aab5/packages/playwright-test/src/runner/testGroups.ts#LL155C1-L156C1

Resulted in less variance among shards and savings of up to a few minutes on the long pole.

**Comment by pavelfeldman** [comment]
What do you sort and by which criteria?

**Comment by pastelsky** [comment]
Very crudely on the test paths for now. We've patched the runner with — 
```
 const sortedTestGroups = testGroups.sort((a, b) => a.requireFile.localeCompare(b.requireFile))
```
for now, but there might be a better way.

I guess an extension of this idea is support custom test sequencers (like Jest), but to start with, predictable slicing based on similar paths might be faster, more stable and more intuitive.

**Comment by pavelfeldman** [comment]
Do you run several projects (browsers) in one run? Your optimization would only help if you are running all the browsers together.

**Comment by dgozman** [comment]
Closing because we lack the information to act on this issue. If you can provide a full repro, please file a new issue by filling the "Bug Report" template, and link to this one.

**Comment by pastelsky** [comment]
@dgozman apologies for the late reply. Sorting the tests has a large impact in two cases —

1. Running with `repeat-each` — Running the tests twice in the same shard rather than in different ones ensures that the test assets only need to be compiled once.
2. Running the same test in multiple projects — Here too the cost of compilation is saved

Playwright's quite fast at running tests in parallel, and for us, the bottleneck really is compiled times (we deal with heavy components). Do you see any problems if these were sorted? If not, I can give this a shot.
